File: packages/sdk-ruby/lib/pocketping/bridges/discord_bridge.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"

module PocketPing
  module Bridge
    # Discord webhook bridge for sending notifications via Discord webhooks
    #
    # @example Basic usage
    #   bridge = DiscordWebhookBridge.new(
    #     webhook_url: "https://discord.com/api/webhooks/123/abc..."
    #   )
    #   pp = PocketPing::Client.new(bridges: [bridge])
    #
    # @example With custom username and avatar
    #   bridge = DiscordWebhookBridge.new(
    #     webhook_url: "https://discord.com/api/webhooks/123/abc...",
    #     username: "PocketPing Bot",
    #     avatar_url: "https://example.com/avatar.png"
    #   )
    class DiscordWebhookBridge < Base
      # @param webhook_url [String] Discord webhook URL
      # @param username [String, nil] Override webhook username
      # @param avatar_url [String, nil] Override webhook avatar URL
      def initialize(webhook_url:, username: nil, avatar_url: nil)
        super()
        @webhook_url = webhook_url
        @username = username
        @avatar_url = avatar_url

        # Extract webhook ID and token from URL for edit/delete operations
        match = webhook_url.match(%r{/webhooks/(\d+)/([^/]+)})
        @webhook_id = match[1] if match
        @webhook_token = match[2] if match
      end

      # @return [String] "discord_webhook"
      def name
        "discord_webhook"
      end

      # Called when a new chat session is created
      #
      # @param session [Session] The new session
      # @return [void]
      def on_new_session(session)
        visitor_id = session.visitor_id || "Unknown"
        url = session.metadata&.url || "No URL"

        content = format_new_session_message(visitor_id, url)
        send_webhook_message(content)
      rescue StandardError => e
        warn "[PocketPing] DiscordWebhookBridge error in on_new_session: #{e.message}"
      end

      # Called when a visitor sends a message
      #
      # @param message [Message] The message
      # @param session [Session] The session
      # @return [BridgeMessageResult, nil] Result with Discord message ID
      def on_visitor_message(message, session)
        visitor_id = session.visitor_id || "Unknown"
        content = message.content || ""

        text = format_visitor_message(visitor_id, content)
        result = send_webhook_message(text, wait: true)
        return nil unless result

        message_id = result["id"]
        return nil unless message_id

        BridgeMessageResult.new(message_id: message_id)
      rescue StandardError => e
        warn "[PocketPing] DiscordWebhookBridge error in on_visitor_message: #{e.message}"
        nil
      end

      # Called when a message is edited
      #
      # @param message [Message] The edited message
      # @param session [Session] The session
      # @param discord_message_id [String] The Discord message ID to edit
      # @return [void]
      def on_message_edit(message, session, discord_message_id)
        return unless discord_message_id && @webhook_id && @webhook_token

        visitor_id = session.visitor_id || "Unknown"
        content = message.content || ""
        text = format_visitor_message(visitor_id, content, edited: true)

        edit_webhook_message(discord_message_id, text)
      rescue StandardError => e
        warn "[PocketPing] DiscordWebhookBridge error in on_message_edit: #{e.message}"
      end

      # Called when a message is deleted
      #
      # @param message [Message] The deleted message
      # @param session [Session] The session
      # @param discord_message_id [String] The Discord message ID to delete
      # @return [void]
      def on_message_delete(message, session, discord_message_id)
        return unless discord_message_id && @webhook_id && @webhook_token

        delete_webhook_message(discord_message_id)
      rescue StandardError => e
        warn "[PocketPing] DiscordWebhookBridge error in on_message_delete: #{e.message}"
      end

      private

      def format_new_session_message(visitor_id, url)
        [
          "\u{1F195} New chat session",
          "\u{1F464} Visitor: #{visitor_id}",
          "\u{1F4CD} #{url}"
        ].join("\n")
      end

      def format_visitor_message(visitor_id, content, edited: false)
        prefix = edited ? "\u{1F4DD} [edited] " : ""
        "\u{1F4AC} #{prefix}#{visitor_id}:\n#{content}"
      end

      def send_webhook_message(content, wait: false)
        uri = URI(@webhook_url)
        uri.query = "wait=true" if wait

        body = { content: content }
        body[:username] = @username if @username
        body[:avatar_url] = @avatar_url if @avatar_url

        make_request(uri, :post, body)
      end

      def edit_webhook_message(message_id, content)
        uri = URI("https://discord.com/api/webhooks/#{@webhook_id}/#{@webhook_token}/messages/#{message_id}")
        body = { content: content }
        make_request(uri, :patch, body)
      end

      def delete_webhook_message(message_id)
        uri = URI("https://discord.com/api/webhooks/#{@webhook_id}/#{@webhook_token}/messages/#{message_id}")
        make_request(uri, :delete, nil)
      end

      def make_request(uri, method, body)
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true
        http.open_timeout = 10
        http.read_timeout = 30

        request = case method
                  when :post
                    Net::HTTP::Post.new(uri)
                  when :patch
                    Net::HTTP::Patch.new(uri)
                  when :delete
                    Net::HTTP::Delete.new(uri)
                  else
                    raise ArgumentError, "Unknown HTTP method: #{method}"
                  end

        request["Content-Type"] = "application/json"
        request.body = body.to_json if body

        response = http.request(request)

        # Discord returns 204 No Content for successful delete
        return {} if response.is_a?(Net::HTTPNoContent)

        unless response.is_a?(Net::HTTPSuccess)
          warn "[PocketPing] Discord API error: #{response.code} - #{response.body}"
          return nil
        end

        return {} if response.body.nil? || response.body.empty?

        JSON.parse(response.body)
      rescue Net::OpenTimeout, Net::ReadTimeout => e
        warn "[PocketPing] Discord API timeout: #{e.message}"
        nil
      rescue JSON::ParserError => e
        warn "[PocketPing] Discord API invalid JSON response: #{e.message}"
        nil
      rescue StandardError => e
        warn "[PocketPing] Discord API error: #{e.message}"
        nil
      end
    end

    # Discord bot bridge for sending notifications via Discord Bot API
    #
    # @example Basic usage
    #   bridge = DiscordBotBridge.new(
    #     bot_token: "Bot MTIz...",
    #     channel_id: "1234567890"
    #   )
    #   pp = PocketPing::Client.new(bridges: [bridge])
    class DiscordBotBridge < Base
      DISCORD_API_BASE = "https://discord.com/api/v10"

      # @param bot_token [String] Discord bot token
      # @param channel_id [String] Channel ID to send messages to
      def initialize(bot_token:, channel_id:)
        super()
        @bot_token = bot_token
        @channel_id = channel_id
      end

      # @return [String] "discord_bot"
      def name
        "discord_bot"
      end

      # Called when a new chat session is created
      #
      # @param session [Session] The new session
      # @return [void]
      def on_new_session(session)
        visitor_id = session.visitor_id || "Unknown"
        url = session.metadata&.url || "No URL"

        content = format_new_session_message(visitor_id, url)
        send_message(content)
      rescue StandardError => e
        warn "[PocketPing] DiscordBotBridge error in on_new_session: #{e.message}"
      end

      # Called when a visitor sends a message
      #
      # @param message [Message] The message
      # @param session [Session] The session
      # @return [BridgeMessageResult, nil] Result with Discord message ID
      def on_visitor_message(message, session)
        visitor_id = session.visitor_id || "Unknown"
        content = message.content || ""

        # Send typing indicator first
        trigger_typing

        text = format_visitor_message(visitor_id, content)
        result = send_message(text)
        return nil unless result

        message_id = result["id"]
        return nil unless message_id

        BridgeMessageResult.new(message_id: message_id)
      rescue StandardError => e
        warn "[PocketPing] DiscordBotBridge error in on_visitor_message: #{e.message}"
        nil
      end

      # Called when a message is edited
      #
      # @param message [Message] The edited message
      # @param session [Session] The session
      # @param discord_message_id [String] The Discord message ID to edit
      # @return [void]
      def on_message_edit(message, session, discord_message_id)
        return unless discord_message_id

        visitor_id = session.visitor_id || "Unknown"
        content = message.content || ""
        text = format_visitor_message(visitor_id, content, edited: true)

        edit_message(discord_message_id, text)
      rescue StandardError => e
        warn "[PocketPing] DiscordBotBridge error in on_message_edit: #{e.message}"
      end

      # Called when a message is deleted
      #
      # @param message [Message] The deleted message
      # @param session [Session] The session
      # @param discord_message_id [String] The Discord message ID to delete
      # @return [void]
      def on_message_delete(message, session, discord_message_id)
        return unless discord_message_id

        delete_message(discord_message_id)
      rescue StandardError => e
        warn "[PocketPing] DiscordBotBridge error in on_message_delete: #{e.message}"
      end

      # Called when visitor starts/stops typing
      #
      # @param session_id [String] The session ID
      # @param is_typing [Boolean] Whether the visitor is typing
      # @return [void]
      def on_typing(session_id, is_typing)
        return unless is_typing

        trigger_typing
      rescue StandardError => e
        warn "[PocketPing] DiscordBotBridge error in on_typing: #{e.message}"
      end

      private

      def format_new_session_message(visitor_id, url)
        [
          "\u{1F195} New chat session",
          "\u{1F464} Visitor: #{visitor_id}",
          "\u{1F4CD} #{url}"
        ].join("\n")
      end

      def format_visitor_message(visitor_id, content, edited: false)
        prefix = edited ? "\u{1F4DD} [edited] " : ""
        "\u{1F4AC} #{prefix}#{visitor_id}:\n#{content}"
      end

      def send_message(content)
        uri = URI("#{DISCORD_API_BASE}/channels/#{@channel_id}/messages")
        body = { content: content }
        make_request(uri, :post, body)
      end

      def edit_message(message_id, content)
        uri = URI("#{DISCORD_API_BASE}/channels/#{@channel_id}/messages/#{message_id}")
        body = { content: content }
        make_request(uri, :patch, body)
      end

      def delete_message(message_id)
        uri = URI("#{DISCORD_API_BASE}/channels/#{@channel_id}/messages/#{message_id}")
        make_request(uri, :delete, nil)
      end

      def trigger_typing
        uri = URI("#{DISCORD_API_BASE}/channels/#{@channel_id}/typing")
        make_request(uri, :post, nil)
      end

      def make_request(uri, method, body)
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true
        http.open_timeout = 10
        http.read_timeout = 30

        request = case method
                  when :post
                    Net::HTTP::Post.new(uri)
                  when :patch
                    Net::HTTP::Patch.new(uri)
                  when :delete
                    Net::HTTP::Delete.new(uri)
                  else
                    raise ArgumentError, "Unknown HTTP method: #{method}"
                  end

        request["Authorization"] = "Bot #{@bot_token}"
        request["Content-Type"] = "application/json"
        request.body = body.to_json if body

        response = http.request(request)

        # Discord returns 204 No Content for successful delete and typing
        return {} if response.is_a?(Net::HTTPNoContent)

        unless response.is_a?(Net::HTTPSuccess)
          warn "[PocketPing] Discord API error: #{response.code} - #{response.body}"
          return nil
        end

        return {} if response.body.nil? || response.body.empty?

        JSON.parse(response.body)
      rescue Net::OpenTimeout, Net::ReadTimeout => e
        warn "[PocketPing] Discord API timeout: #{e.message}"
        nil
      rescue JSON::ParserError => e
        warn "[PocketPing] Discord API invalid JSON response: #{e.message}"
        nil
      rescue StandardError => e
        warn "[PocketPing] Discord API error: #{e.message}"
        nil
      end
    end
  end
end

File: packages/sdk-go/discord_bridge.go
package pocketping

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"strings"
	"time"
)

// DiscordWebhookBridge sends notifications to Discord via webhook.
type DiscordWebhookBridge struct {
	BaseBridge
	WebhookURL string
	Username   string
	AvatarURL  string

	httpClient *http.Client
	pp         *PocketPing
}

// DiscordWebhookOption is a functional option for DiscordWebhookBridge.
type DiscordWebhookOption func(*DiscordWebhookBridge)

// WithDiscordWebhookUsername sets the webhook username.
func WithDiscordWebhookUsername(username string) DiscordWebhookOption {
	return func(d *DiscordWebhookBridge) {
		d.Username = username
	}
}

// WithDiscordWebhookAvatarURL sets the webhook avatar URL.
func WithDiscordWebhookAvatarURL(avatarURL string) DiscordWebhookOption {
	return func(d *DiscordWebhookBridge) {
		d.AvatarURL = avatarURL
	}
}

// WithDiscordWebhookHTTPClient sets a custom HTTP client.
func WithDiscordWebhookHTTPClient(client *http.Client) DiscordWebhookOption {
	return func(d *DiscordWebhookBridge) {
		d.httpClient = client
	}
}

// NewDiscordWebhookBridge creates a new Discord webhook bridge.
func NewDiscordWebhookBridge(webhookURL string, opts ...DiscordWebhookOption) *DiscordWebhookBridge {
	d := &DiscordWebhookBridge{
		BaseBridge: BaseBridge{BridgeName: "discord-webhook"},
		WebhookURL: webhookURL,
		Username:   "PocketPing",
		httpClient: &http.Client{Timeout: 30 * time.Second},
	}

	for _, opt := range opts {
		opt(d)
	}

	return d
}

// Init initializes the Discord webhook bridge.
func (d *DiscordWebhookBridge) Init(ctx context.Context, pp *PocketPing) error {
	d.pp = pp
	return nil
}

// OnNewSession sends a notification when a new session is created.
func (d *DiscordWebhookBridge) OnNewSession(ctx context.Context, session *Session) error {
	visitorName := d.getVisitorName(session)
	pageURL := ""
	if session.Metadata != nil && session.Metadata.URL != "" {
		pageURL = session.Metadata.URL
	}

	content := fmt.Sprintf("🆕 New chat session\n👤 Visitor: %s", visitorName)
	if pageURL != "" {
		content += fmt.Sprintf("\n📍 %s", pageURL)
	}

	_, err := d.sendWebhookMessage(ctx, content, "")
	if err != nil {
		log.Printf("[DiscordWebhookBridge] OnNewSession error: %v", err)
	}
	return nil
}

// OnVisitorMessage sends a notification when a visitor sends a message.
func (d *DiscordWebhookBridge) OnVisitorMessage(ctx context.Context, message *Message, session *Session) error {
	visitorName := d.getVisitorName(session)
	content := fmt.Sprintf("💬 %s:\n%s", visitorName, message.Content)

	// Note: Discord webhooks don't return message IDs in a way that allows editing
	// For full edit/delete support, use DiscordBotBridge instead
	var replyToMessageID string
	if message.ReplyTo != "" && d.pp != nil {
		if storage, ok := d.pp.GetStorage().(StorageWithBridgeIDs); ok {
			bridgeIDs, err := storage.GetBridgeMessageIDs(ctx, message.ReplyTo)
			if err == nil && bridgeIDs != nil && bridgeIDs.DiscordMessageID != "" {
				replyToMessageID = bridgeIDs.DiscordMessageID
			}
		}
	}

	result, err := d.sendWebhookMessage(ctx, content, replyToMessageID)
	if err != nil {
		log.Printf("[DiscordWebhookBridge] OnVisitorMessage error: %v", err)
		return nil
	}

	if result != nil && result.DiscordMessageID != "" && d.pp != nil {
		if storage, ok := d.pp.GetStorage().(StorageWithBridgeIDs); ok {
			_ = storage.SaveBridgeMessageIDs(ctx, message.ID, BridgeMessageIds{
				DiscordMessageID: result.DiscordMessageID,
			})
		}
	}
	return nil
}

// OnOperatorMessage is called when an operator sends a message.
func (d *DiscordWebhookBridge) OnOperatorMessage(ctx context.Context, message *Message, session *Session, sourceBridge string, operatorName string) error {
	if sourceBridge == d.Name() {
		return nil
	}

	name := operatorName
	if name == "" {
		name = "Operator"
	}

	content := fmt.Sprintf("👨‍💼 %s:\n%s", name, message.Content)

	_, err := d.sendWebhookMessage(ctx, content, "")
	if err != nil {
		log.Printf("[DiscordWebhookBridge] OnOperatorMessage error: %v", err)
	}
	return nil
}

// OnTyping is called when visitor starts/stops typing.
func (d *DiscordWebhookBridge) OnTyping(ctx context.Context, sessionID string, isTyping bool) error {
	// Discord webhooks don't support typing indicators
	return nil
}

// OnMessageRead is called when messages are marked as read.
func (d *DiscordWebhookBridge) OnMessageRead(ctx context.Context, sessionID string, messageIDs []string, status MessageStatus) error {
	return nil
}

// OnCustomEvent is called when a custom event is triggered.
func (d *DiscordWebhookBridge) OnCustomEvent(ctx context.Context, event CustomEvent, session *Session) error {
	visitorName := d.getVisitorName(session)
	content := fmt.Sprintf("📌 Event from %s: %s", visitorName, event.Name)

	if len(event.Data) > 0 {
		dataJSON, err := json.Marshal(event.Data)
		if err == nil {
			content += fmt.Sprintf("\n📦 %s", string(dataJSON))
		}
	}

	_, err := d.sendWebhookMessage(ctx, content, "")
	if err != nil {
		log.Printf("[DiscordWebhookBridge] OnCustomEvent error: %v", err)
	}
	return nil
}

// OnIdentityUpdate is called when a user identifies themselves.
func (d *DiscordWebhookBridge) OnIdentityUpdate(ctx context.Context, session *Session) error {
	if session.Identity == nil {
		return nil
	}

	content := fmt.Sprintf("🔐 User identified\n👤 ID: %s", session.Identity.ID)
	if session.Identity.Name != "" {
		content += fmt.Sprintf("\n📛 Name: %s", session.Identity.Name)
	}
	if session.Identity.Email != "" {
		content += fmt.Sprintf("\n📧 Email: %s", session.Identity.Email)
	}

	_, err := d.sendWebhookMessage(ctx, content, "")
	if err != nil {
		log.Printf("[DiscordWebhookBridge] OnIdentityUpdate error: %v", err)
	}
	return nil
}

type discordWebhookPayload struct {
	Content          string                   `json:"content"`
	Username         string                   `json:"username,omitempty"`
	AvatarURL        string                   `json:"avatar_url,omitempty"`
	MessageReference *discordMessageReference `json:"message_reference,omitempty"`
}

type discordMessageReference struct {
	MessageID string `json:"message_id"`
}

func (d *DiscordWebhookBridge) sendWebhookMessage(ctx context.Context, content string, replyToMessageID string) (*BridgeMessageResult, error) {
	payload := discordWebhookPayload{
		Content:   content,
		Username:  d.Username,
		AvatarURL: d.AvatarURL,
	}
	if replyToMessageID != "" {
		payload.MessageReference = &discordMessageReference{MessageID: replyToMessageID}
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("marshal payload: %w", err)
	}

	// Add ?wait=true to get the message ID back
	webhookURL := d.WebhookURL
	if !strings.Contains(webhookURL, "?") {
		webhookURL += "?wait=true"
	} else {
		webhookURL += "&wait=true"
	}

	req, err := http.NewRequestWithContext(ctx, "POST", webhookURL, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := d.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		respBody, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("discord error: %s (status %d)", string(respBody), resp.StatusCode)
	}

	// Parse response to get message ID
	var discordResp struct {
		ID string `json:"id"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&discordResp); err == nil && discordResp.ID != "" {
		return &BridgeMessageResult{
			DiscordMessageID: discordResp.ID,
		}, nil
	}

	return nil, nil
}

func (d *DiscordWebhookBridge) getVisitorName(session *Session) string {
	if session.Identity != nil && session.Identity.Name != "" {
		return session.Identity.Name
	}
	if session.Identity != nil && session.Identity.Email != "" {
		return session.Identity.Email
	}
	return session.VisitorID
}

// Ensure DiscordWebhookBridge implements Bridge interface
var _ Bridge = (*DiscordWebhookBridge)(nil)

// DiscordBotBridge sends notifications to Discord using a bot token.
// This supports full edit/delete functionality.
type DiscordBotBridge struct {
	BaseBridge
	BotToken  string
	ChannelID string

	httpClient *http.Client
	pp         *PocketPing
}

// DiscordBotOption is a functional option for DiscordBotBridge.
type DiscordBotOption func(*DiscordBotBridge)

// WithDiscordBotHTTPClient sets a custom HTTP client.
func WithDiscordBotHTTPClient(client *http.Client) DiscordBotOption {
	return func(d *DiscordBotBridge) {
		d.httpClient = client
	}
}

// NewDiscordBotBridge creates a new Discord bot bridge.
func NewDiscordBotBridge(botToken, channelID string, opts ...DiscordBotOption) *DiscordBotBridge {
	d := &DiscordBotBridge{
		BaseBridge: BaseBridge{BridgeName: "discord-bot"},
		BotToken:   botToken,
		ChannelID:  channelID,
		httpClient: &http.Client{Timeout: 30 * time.Second},
	}

	for _, opt := range opts {
		opt(d)
	}

	return d
}

// Init initializes the Discord bot bridge.
func (d *DiscordBotBridge) Init(ctx context.Context, pp *PocketPing) error {
	d.pp = pp
	return nil
}

// OnNewSession sends a notification when a new session is created.
func (d *DiscordBotBridge) OnNewSession(ctx context.Context, session *Session) error {
	visitorName := d.getVisitorName(session)
	pageURL := ""
	if session.Metadata != nil && session.Metadata.URL != "" {
		pageURL = session.Metadata.URL
	}

	content := fmt.Sprintf("🆕 New chat session\n👤 Visitor: %s", visitorName)
	if pageURL != "" {
		content += fmt.Sprintf("\n📍 %s", pageURL)
	}

	_, err := d.sendMessage(ctx, content, "")
	if err != nil {
		log.Printf("[DiscordBotBridge] OnNewSession error: %v", err)
	}
	return nil
}

// OnVisitorMessage sends a notification when a visitor sends a message.
func (d *DiscordBotBridge) OnVisitorMessage(ctx context.Context, message *Message, session *Session) error {
	visitorName := d.getVisitorName(session)
	content := fmt.Sprintf("💬 %s:\n%s", visitorName, message.Content)

	var replyToMessageID string
	if message.ReplyTo != "" && d.pp != nil {
		if storage, ok := d.pp.GetStorage().(StorageWithBridgeIDs); ok {
			bridgeIDs, err := storage.GetBridgeMessageIDs(ctx, message.ReplyTo)
			if err == nil && bridgeIDs != nil && bridgeIDs.DiscordMessageID != "" {
				replyToMessageID = bridgeIDs.DiscordMessageID
			}
		}
	}

	result, err := d.sendMessage(ctx, content, replyToMessageID)
	if err != nil {
		log.Printf("[DiscordBotBridge] OnVisitorMessage error: %v", err)
		return nil
	}

	// Save bridge message ID for edit/delete support
	if result != nil && result.DiscordMessageID != "" && d.pp != nil {
		if storage, ok := d.pp.GetStorage().(StorageWithBridgeIDs); ok {
			_ = storage.SaveBridgeMessageIDs(ctx, message.ID, BridgeMessageIds{
				DiscordMessageID: result.DiscordMessageID,
			})
		}
	}

	return nil
}

// OnOperatorMessage is called when an operator sends a message.
func (d *DiscordBotBridge) OnOperatorMessage(ctx context.Context, message *Message, session *Session, sourceBridge string, operatorName string) error {
	if sourceBridge == d.Name() {
		return nil
	}

	name := operatorName
	if name == "" {
		name = "Operator"
	}

	content := fmt.Sprintf("👨‍💼 %s:\n%s", name, message.Content)

	_, err := d.sendMessage(ctx, content, "")
	if err != nil {
		log.Printf("[DiscordBotBridge] OnOperatorMessage error: %v", err)
	}
	return nil
}

// OnTyping sends a typing indicator.
func (d *DiscordBotBridge) OnTyping(ctx context.Context, sessionID string, isTyping bool) error {
	if !isTyping {
		return nil
	}

	err := d.triggerTyping(ctx)
	if err != nil {
		log.Printf("[DiscordBotBridge] OnTyping error: %v", err)
	}
	return nil
}

// OnMessageRead is called when messages are marked as read.
func (d *DiscordBotBridge) OnMessageRead(ctx context.Context, sessionID string, messageIDs []string, status MessageStatus) error {
	return nil
}

// OnCustomEvent is called when a custom event is triggered.
func (d *DiscordBotBridge) OnCustomEvent(ctx context.Context, event CustomEvent, session *Session) error {
	visitorName := d.getVisitorName(session)
	content := fmt.Sprintf("📌 Event from %s: %s", visitorName, event.Name)

	if len(event.Data) > 0 {
		dataJSON, err := json.Marshal(event.Data)
		if err == nil {
			content += fmt.Sprintf("\n📦 %s", string(dataJSON))
		}
	}

	_, err := d.sendMessage(ctx, content, "")
	if err != nil {
		log.Printf("[DiscordBotBridge] OnCustomEvent error: %v", err)
	}
	return nil
}

// OnIdentityUpdate is called when a user identifies themselves.
func (d *DiscordBotBridge) OnIdentityUpdate(ctx context.Context, session *Session) error {
	if session.Identity == nil {
		return nil
	}

	content := fmt.Sprintf("🔐 User identified\n👤 ID: %s", session.Identity.ID)
	if session.Identity.Name != "" {
		content += fmt.Sprintf("\n📛 Name: %s", session.Identity.Name)
	}
	if session.Identity.Email != "" {
		content += fmt.Sprintf("\n📧 Email: %s", session.Identity.Email)
	}

	_, err := d.sendMessage(ctx, content, "")
	if err != nil {
		log.Printf("[DiscordBotBridge] OnIdentityUpdate error: %v", err)
	}
	return nil
}

// OnMessageEdit handles message edits.
func (d *DiscordBotBridge) OnMessageEdit(ctx context.Context, sessionID, messageID, content string, editedAt time.Time) (*BridgeMessageResult, error) {
	if d.pp == nil {
		return nil, nil
	}

	storage, ok := d.pp.GetStorage().(StorageWithBridgeIDs)
	if !ok {
		return nil, nil
	}

	bridgeIDs, err := storage.GetBridgeMessageIDs(ctx, messageID)
	if err != nil || bridgeIDs == nil || bridgeIDs.DiscordMessageID == "" {
		return nil, nil
	}

	err = d.editMessage(ctx, bridgeIDs.DiscordMessageID, content+" (edited)")
	if err != nil {
		log.Printf("[DiscordBotBridge] OnMessageEdit error: %v", err)
		return nil, nil
	}

	return &BridgeMessageResult{
		DiscordMessageID: bridgeIDs.DiscordMessageID,
	}, nil
}

// OnMessageDelete handles message deletions.
func (d *DiscordBotBridge) OnMessageDelete(ctx context.Context, sessionID, messageID string, deletedAt time.Time) error {
	if d.pp == nil {
		return nil
	}

	storage, ok := d.pp.GetStorage().(StorageWithBridgeIDs)
	if !ok {
		return nil
	}

	bridgeIDs, err := storage.GetBridgeMessageIDs(ctx, messageID)
	if err != nil || bridgeIDs == nil || bridgeIDs.DiscordMessageID == "" {
		return nil
	}

	err = d.deleteMessage(ctx, bridgeIDs.DiscordMessageID)
	if err != nil {
		log.Printf("[DiscordBotBridge] OnMessageDelete error: %v", err)
	}
	return nil
}

// Discord API helpers

const discordAPIBase = "https://discord.com/api/v10"

type discordMessagePayload struct {
	Content          string                   `json:"content"`
	MessageReference *discordMessageReference `json:"message_reference,omitempty"`
}

type discordMessage struct {
	ID string `json:"id"`
}

func (d *DiscordBotBridge) sendMessage(ctx context.Context, content string, replyToMessageID string) (*BridgeMessageResult, error) {
	apiURL := fmt.Sprintf("%s/channels/%s/messages", discordAPIBase, d.ChannelID)

	payload := discordMessagePayload{Content: content}
	if replyToMessageID != "" {
		payload.MessageReference = &discordMessageReference{MessageID: replyToMessageID}
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("marshal payload: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", apiURL, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bot "+d.BotToken)

	resp, err := d.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("send request: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response: %w", err)
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("discord error: %s (status %d)", string(respBody), resp.StatusCode)
	}

	var msg discordMessage
	if err := json.Unmarshal(respBody, &msg); err != nil {
		return nil, fmt.Errorf("parse response: %w", err)
	}

	return &BridgeMessageResult{
		DiscordMessageID: msg.ID,
	}, nil
}

func (d *DiscordBotBridge) editMessage(ctx context.Context, messageID, content string) error {
	apiURL := fmt.Sprintf("%s/channels/%s/messages/%s", discordAPIBase, d.ChannelID, messageID)

	payload := discordMessagePayload{Content: content}
	body, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("marshal payload: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "PATCH", apiURL, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bot "+d.BotToken)

	resp, err := d.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		respBody, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("discord error: %s (status %d)", string(respBody), resp.StatusCode)
	}

	return nil
}

func (d *DiscordBotBridge) deleteMessage(ctx context.Context, messageID string) error {
	apiURL := fmt.Sprintf("%s/channels/%s/messages/%s", discordAPIBase, d.ChannelID, messageID)

	req, err := http.NewRequestWithContext(ctx, "DELETE", apiURL, nil)
	if err != nil {
		return fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Authorization", "Bot "+d.BotToken)

	resp, err := d.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		respBody, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("discord error: %s (status %d)", string(respBody), resp.StatusCode)
	}

	return nil
}

func (d *DiscordBotBridge) triggerTyping(ctx context.Context) error {
	apiURL := fmt.Sprintf("%s/channels/%s/typing", discordAPIBase, d.ChannelID)

	req, err := http.NewRequestWithContext(ctx, "POST", apiURL, nil)
	if err != nil {
		return fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Authorization", "Bot "+d.BotToken)

	resp, err := d.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("send request: %w", err)
	}
	defer resp.Body.Close()

	return nil
}

func (d *DiscordBotBridge) getVisitorName(session *Session) string {
	if session.Identity != nil && session.Identity.Name != "" {
		return session.Identity.Name
	}
	if session.Identity != nil && session.Identity.Email != "" {
		return session.Identity.Email
	}
	return session.VisitorID
}

// Ensure DiscordBotBridge implements Bridge interface
var _ Bridge = (*DiscordBotBridge)(nil)

// Ensure DiscordBotBridge implements BridgeWithEditDelete interface
var _ BridgeWithEditDelete = (*DiscordBotBridge)(nil)
